### PR TITLE
Add relativistic pair production

### PR DESCRIPTION
**Pull request by amandalund**
This extends the Bethe-Heitler interactor so it's valid at high energies. Changes include:
 - Accounting for the LPM effect in the Bethe-Heitler cross section. The relativistic pair production model samples secondaries in the same way as the Bethe-Heitler model, so they are otherwise identical.
 - A couple small fixes and additional documentation in the Bethe-Heitler interactor
 - Increasing the bremsstrahlung model energy limit to 100 TeV
 - Separating the calculation of the LPM functions from the relativistic bremsstrahlung implementation so it can be used for both pair production and bremsstrahlung
 - Calculating the LPM data on the fly instead of precomputing and storing at initialization.

The changes to the bremsstrahlung tests are just from calculating G(s) and phi(s) on the fly instead of interpolating tabulated data.

### CODE REVIEW

File: src/physics/em/detail/LPMCalculator.hh [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        // Include a dielectric suppression effect in \f$ s \f$ according to
+        // Migdal by multiplying \f$ s \f$ by \f$ 1 + \frac{k_p^2}{k^2} \f$,
+        // where the characteristic photon energy scale \f$ k_p \f$ is defined
+        // in terms of the plasma frequency of the medium \f$ \omega_p \f$: \f$
+        // k_p = \hbar \omega_p \frac{E}{m_e c^2} \f$
+        const real_type k_p = electron_density_ * migdal_constant()
+                              * ipow<2>(epsilon * gamma_energy_);
+        s *= (1 + k_p / ipow<2>(gamma_energy_));
+
+        // Recalculate \f$ \xi \$ from the modified suppression variable
```
Comment (paulromano) [inline src/physics/em/detail/LPMCalculator.hh:150]: Modify comment indicating Eq. (16) from Stanev

File: src/physics/em/detail/LPMCalculator.hh [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+{
+    // Suppression variable \f$ s' \f$. For bremsstrahlung \f$ s' =
+    // \sqrt{\frac{E_\textrm{LPM} k}{8 E (E - k)}} \f$, and for pair production
+    // \f$ s' = \sqrt{\frac{E_\textrm{LPM} k}{8 E (k - E)}} \f$, where \f$ k \$
+    // is the photon energy and \f$ E \f$ is the electon (or positron) energy
+    const real_type s_prime = std::sqrt(
+        lpm_energy_ / (8 * epsilon * gamma_energy_ * std::fabs(epsilon - 1)));
+
+    // TODO: In the Geant4 relativistic pair production model the denominator
+    // is 184 instead of 184.15 -- why? Will it matter?
```
Comment (paulromano) [inline src/physics/em/detail/LPMCalculator.hh:122]: Some references do indeed have 184 (e.g., [Klein](https://doi.org/10.1103/RevModPhys.71.1501)), although the original calculation by [Tsai](https://doi.org/10.1103/RevModPhys.46.815) was 184.15. It looks like this was all based on some numerical approximations and various models, so there shouldn't be any practical difference between 184 and 184.15.

File: src/physics/em/detail/LPMCalculator.hh [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    const real_type s1 = ipow<2>(element_.cbrt_z() / real_type(184.15));
+
+    // Calculate \f$ \xi(s') \f$ and \f$ s = \frac{s'}{\sqrt{\xi(s')}} \f$
```
Comment (paulromano) [inline src/physics/em/detail/LPMCalculator.hh:125]: Modify comment mentioning Eq. (21) in Stanev

**Review by paulromano — CHANGES_REQUESTED** [review]
Nice work @amandalund! Only a few small requests to improve comments here and there

File: src/physics/em/detail/RBDiffXsCalculator.hh [hunk 4]
```diff
@@ -129,20 +120,22 @@ class RBDiffXsCalculator
...
                     + value_as<units::MevMass>(shared.electron_mass))
 {
     real_type density_factor = material.electron_density() * migdal_constant();
     density_corr_            = density_factor * ipow<2>(total_energy_);
 
-    lpm_energy_ = material.radiation_length()
-                  * value_as<MevPerCm>(lpm_constant());
-    real_type lpm_threshold = lpm_energy_ * std::sqrt(density_factor);
+    real_type lpm_energy = material.radiation_length()
+                           * value_as<MevPerCm>(lpm_constant());
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/physics/em/detail/RBDiffXsCalculator.hh:137]: 4 pi  hbar  c = 2 h c

[the remark this answers is not in the document]
Comment (paulromano) [reply src/physics/em/detail/RBDiffXsCalculator.hh:137]: Right -- in the G4 PRM section 10.2.2, they have 4hc in the denominator which is why I was confused

File: src/physics/em/detail/LPMCalculator.hh [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+    if (dielectric_suppression_)
+    {
+        // Include a dielectric suppression effect in \f$ s \f$ according to
+        // Migdal by multiplying \f$ s \f$ by \f$ 1 + \frac{k_p^2}{k^2} \f$,
+        // where the characteristic photon energy scale \f$ k_p \f$ is defined
+        // in terms of the plasma frequency of the medium \f$ \omega_p \f$: \f$
+        // k_p = \hbar \omega_p \frac{E}{m_e c^2} \f$
+        const real_type k_p = electron_density_ * migdal_constant()
+                              * ipow<2>(epsilon * gamma_energy_);
```
[the remark this answers is not in the document]
Comment (paulromano) [reply src/physics/em/detail/LPMCalculator.hh:147]: G4 PRM uses k<sub>p</sub> for this term, as do other references, so I think that part is fine. Just a matter of whether it is squared or not

**Comment by whokion** [comment]
@amandalund nice work! - also good that you factor out the LPM calculation from the RB model and reuse it  inboth places.  Given that this MR fixes the energy response in the tail of TestEM3, we should merge this rather quickly and update the white paper.

File: src/physics/em/detail/LPMCalculator.hh (hunk 5 above)
[the remark this answers is not in the document]
Comment (amandalund) [reply src/physics/em/detail/LPMCalculator.hh:147]: Yes you're right -- updated.

File: src/physics/em/BetheHeitlerModel.hh [hunk 6]
```diff
@@ -22,7 +22,9 @@ class BetheHeitlerModel final : public Model
 {
   public:
     // Construct from model ID and other necessary data
-    BetheHeitlerModel(ModelId id, const ParticleParams& particles);
+    BetheHeitlerModel(ModelId               id,
+                      const ParticleParams& particles,
+                      bool                  enable_lpm = true);
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/physics/em/BetheHeitlerModel.hh:27]: Done!

File: src/physics/em/detail/LPMCalculator.hh [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+            {
+                result.g = std::tanh(R(-0.160723) + R(3.755030) * s
+                                     - R(1.798138) * s2 + R(0.672827) * s3
+                                     - R(0.120772) * s4);
+            }
+            else
+            {
+                result.g = 1 - R(0.0230655) / s4;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/physics/em/detail/LPMCalculator.hh:241]: Nice, thanks!

**Comment by amandalund** [comment]
Thanks @paulromano, @whokion, @sethrj (or Kali?)! I think I've addressed all your comments.

**Review by sethrj — APPROVED** [review]
Nice work @amandalund , especially with improving agreement with geant4!

**Review by paulromano — APPROVED** [review]

**Review by whokion — APPROVED** [review]

**Pull request merged**